### PR TITLE
Add place types

### PR DESCRIPTION
**Pull request by nicksellen**
This is basically the same as https://github.com/yunity/karrot-backend/pull/1122 but without changing anything about the place statuses. The aim is to reduce the complexity, so I can get something mergable sooner, and deal with any changes to the handling of statuses after that.

It adds a new model "place types" so we can have different types of place with their own icons (and in the future, perhaps behaviour).

There are more detailed thoughts about it on the community forum post --> https://community.foodsaving.world/t/custom-place-types/705/4

## TODO

- [x] vaguely check if the existing frontend will work out with this backend
- [x] decide on initial place types

### CODE REVIEW

File: karrot/places/tests/test_place_types_api.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        self.client.force_login(user=self.member)
+        response = self.client.delete(f'/api/place-types/{self.place_type.id}/')
+        self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT, response.data)
+
+    def test_cannot_delete_with_places(self):
+        self.client.force_login(user=self.member)
+        place = PlaceFactory(place_type=self.place_type)
+        response = self.client.delete(f'/api/place-types/{self.place_type.id}/')
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN, response.data)
+        # make sure we can delete it if we get rid of the activity
```
Comment (tiltec) [inline karrot/places/tests/test_place_types_api.py:99]: typo: activity -> place

Comment (tiltec) [reply karrot/places/tests/test_place_types_api.py:99]: Although places cannot be deleted, just archived! So I wonder if it actually makes sense to delete place types, except newly created ones.

File: karrot/places/tests/test_models.py [hunk 2]
```diff
@@ -10,22 +10,26 @@
...
+        Place.objects.create(name='abcdef', **self.defaults)
         with self.assertRaises(IntegrityError):
-            Place.objects.create(name='abcdef', group=self.group)
+            Place.objects.create(name='abcdef', **self.defaults)
 
     def test_create_place_with_same_name_in_different_groups_works(self):
-        Place.objects.create(name='abcdef', group=self.group)
-        Place.objects.create(name='abcdef', group=GroupFactory())
+        Place.objects.create(name='abcdef', **self.defaults)
+        Place.objects.create(name='abcdef', **{**self.defaults, 'group': GroupFactory()})
```
Comment (tiltec) [inline karrot/places/tests/test_models.py:29]: next level unpacking ;)

File: karrot/places/models.py [hunk 3]
```diff
@@ -13,6 +14,32 @@ class PlaceStatus(models.TextChoices):
...
+    ARCHIVED = 'archived'
+
+
+class PlaceType(BaseModel, UpdatedAtMixin):
+    DEFAULT_STATUS = PlaceTypeStatus.ACTIVE.value
+
+    group = models.ForeignKey('groups.Group', on_delete=models.CASCADE, related_name='place_types')
+    name = models.CharField(max_length=80)
+    name_is_translatable = models.BooleanField(default=True)
+    icon = models.CharField(max_length=32)
```
Comment (tiltec) [inline karrot/places/models.py:28]: Maybe 32 characters can be too short for some icons? Just had a look, there's `fas fa-american-sign-language-interpreting` in https://quasarframework.github.io/quasar-ui-qiconpicker/icons, so we might just push max_length to 100 to be safe.

(I guess that also affects `ActivityType`?)

**Review by tiltec — APPROVED** [review]
Seems good! I went over it and wrote some thoughts down, maybe it's helpful to you.

File: karrot/places/filters.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+from django_filters import rest_framework as filters
+
+from karrot.places.models import PlaceType
+
+
+class PlaceTypeFilter(filters.FilterSet):
+    group = filters.NumberFilter(field_name='group')
```
[the remark this answers is not in the document]
Comment (nicksellen) [reply karrot/places/filters.py:7]: Nope, seems an improvement, also changing it for activity types :+1:

File: karrot/places/permissions.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+            return not obj.places.exists()
+        return True
+
+
+class CannotChangeGroup(permissions.BasePermission):
+    message = 'You cannot change the group for a type'
+
+    def has_object_permission(self, request, view, obj):
+        if view.action == 'partial_update':
+            return 'group' not in request.data
```
[the remark this answers is not in the document]
Comment (nicksellen) [reply karrot/places/permissions.py:20]: I'd prefer it more explicit, shouldn't even _try_ to change the group. Encodes the intention better imo.

File: karrot/places/permissions.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        return True
+
+
+class IsGroupEditor(permissions.BasePermission):
+    message = 'You need to be a group editor'
+
+    def has_permission(self, request, view):
+        if view.action == 'create':
+            if 'group' in request.data:
+                group = Group.objects.filter(id=request.data['group'], members=request.user).first()
```
[the remark this answers is not in the document]
Comment (nicksellen) [reply karrot/places/permissions.py:30]: Yup, good idea. Moving the checks over there (also for activities permissions).

File: karrot/places/serializers.py [hunk 7]
```diff
@@ -48,6 +120,11 @@ class Meta:
     status = serializers.ChoiceField(choices=PlaceStatus.choices, default=PlaceModel.DEFAULT_STATUS)
     is_subscribed = serializers.SerializerMethodField()
 
+    def __init__(self, *args, **kwargs):
+        super(PlaceSerializer, self).__init__(*args, **kwargs)
+        # override the field definition so we can provide a default option
+        self.fields['place_type'].required = False
+
```
[the remark this answers is not in the document]
Comment (nicksellen) [reply karrot/places/serializers.py:127]: Much neater, thanks!

File: karrot/places/tests/test_models.py (hunk 2 above)
Comment (nicksellen) [reply karrot/places/tests/test_models.py:29]: ** ‿ **

File: karrot/places/tests/test_place_types_api.py (hunk 1 above)
Comment (nicksellen) [reply karrot/places/tests/test_place_types_api.py:99]: Yup, maybe doesn't actually get used in the API, although I would just keep it possible.

File: karrot/places/models.py (hunk 3 above)
Comment (nicksellen) [reply karrot/places/models.py:28]: Yup, will switch to 100.

File: karrot/places/filters.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nicksellen) [reply karrot/places/filters.py]: Fixed in https://github.com/yunity/karrot-backend/pull/1159/commits/6fd519773f4b27c229871ccce1165749837747e1

File: karrot/places/permissions.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nicksellen) [reply karrot/places/permissions.py]: Fixed in https://github.com/yunity/karrot-backend/pull/1159/commits/4b46d5a38315fa14b8909e0ba16bd2bb4b7e61f2

File: karrot/places/serializers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nicksellen) [reply karrot/places/serializers.py]: Fixed in https://github.com/yunity/karrot-backend/pull/1159/commits/bebe346bb75e9191a237c37e03ea5300bcb03c6e

File: karrot/places/models.py (hunk 3 above)
Comment (nicksellen) [reply karrot/places/models.py]: Fixed in https://github.com/yunity/karrot-backend/pull/1159/commits/d3b2d39f70d704552284c77db9429dfc83e3effa

File: karrot/places/serializers.py [hunk 8]
```diff
@@ -71,6 +148,14 @@ def create(self, validated_data):
         place.group.refresh_active_status()
         return place
 
+    def validate(self, attrs):
+        if not self.instance and not attrs.get('place_type'):
+            """creating place without place type, we'll provide a default"""
+            group = attrs.get('group')
+            place_type_name = next(iter(default_place_types))
```
[the remark this answers is not in the document]
Comment (nicksellen) [reply karrot/places/serializers.py]: Fixed in https://github.com/yunity/karrot-backend/pull/1159/commits/49cdb709449675579880672295286ad52ad5ca17

**Comment by nicksellen** [comment]
Yup, very helpful, thanks :star: 

Let's hope the tests are good enough to properly test the code changes from the review...

**Pull request merged**